homeassistant/components/light/flux_led.py

```diff
@@ -135,8 +143,83 @@ def supported_features(self):
+
+        # # Cancel scheduled callback of fading action.
+        # if self.transition_stop_handle:
+        #     self.transition_stop_handle.cancel()
+
+        # If light was previously faded to black, restore previous color.
+        if self.previous_brightness or self.previous_color_rgb:
+            kwargs[ATTR_BRIGHTNESS] = self.previous_brightness
+            kwargs[ATTR_RGB_COLOR] = self.previous_color_rgb
+            self.previous_brightness = None
+            self.previous_color_rgb = None
+
+        if ATTR_TRANSITION in kwargs:
+            # if light is not on assume the user wants
+            # to transition from black to the new color
+            if not self.is_on:
+                self._bulb.setRgb(0, 0, 0)
+                self._bulb.turnOn()
+
+            self.set_color_transition(**kwargs)
+        else:
+            if not self.is_on:
+                self._bulb.turnOn()
+
+            self.set_color(**kwargs)
+
+            # notify UI of change when turned on without transition
+            self.schedule_update_ha_state(force_refresh=True)
+
+    def set_color_transition(self, **kwargs):
+        """Set a color transition for light."""
+
+        from flux_led import utils
+
+        # flux_led support integer delay from 1-30 seconds
+        delay_seconds = int(kwargs[ATTR_TRANSITION])
+        transition_time = utils.delayToSpeed(delay_seconds)
+
+        to_rgb = kwargs.get(ATTR_RGB_COLOR)
+        transition_type = 'gradual'
+
+        # Flux led support transitioning between 16 colors and repeats
+        # the cycle after completion. Fading starts with the current color.
+        # This fills the entire transition cycle with the 'to' color. This
+        # gives us enough time to execute a transition stop after a delay.
+        color_list = [to_rgb] * 16
+
+        # Function to stop transition by turning light on to desired color.
+        kwargs_no_transition = kwargs.copy()
+        del kwargs_no_transition['transition']
+        # stop_transition = partial(self.turn_on, **kwargs_no_transition)
+
+        # # Schedule a callback to turn on the bulb at the requested color,
+        # # effectively cancelling the custom color pattern. Since the color
+        # # pattern continuously cycles between 16 of the same color this is
+        # # strictly not needed, but it the nice thing to do
+        # self.transition_stop_handle = self.hass.loop.call_later(
+        #     delay_seconds,
+        #     stop_transition
+        # )
+
+        # Set a custom color pattern function into the led controller and
+        # execute it.
+        self._bulb.setCustomPattern(
+            color_list,
+            transition_time,
+            transition_type
+        )
+
+        # Wait for transition to finish and call turn_on again without
+        # transition argument to stop the custom pattern function as set
+        # the current state of the entity.
+        time.sleep(delay_seconds)
```

Comment:
I was using a asyncio call_later before (see commented out code) to schedule the call but I ran into the issue that the state of the entity would not update after the transition ended (only later at the update interval when it would poll the state (refresh)).

To my understanding whenever a service call is made to turn on/off a light it calls an entity update implicitly after the entity turn_on/turn_off method has returned. So the turn_on would have to wait with returning until the transition is finished (or we should add a entity update to the scheduled task). 

I don't know much about the non asyncio implementation details as all my components have been using asyncio so far. But doesn't a non-asyncio component run in a thread or use a different method to prevent being time.sleep an issue?

If you could shed some light on this that would be nice. It could also be that the behavior I experienced was due to a different cause than what I suspected. Since the fluxled uses UDP and does not always seem to take the state it has been sent.